test/e2e/common/node/pod_admission.go

```diff
@@ -65,6 +66,49 @@ var _ = SIGDescribe("PodOSRejection", framework.WithNodeConformance(), func() {
 	})
 })
 
+var _ = SIGDescribe("PodOSRejectionStatus", func() {
+	f := framework.NewDefaultFramework("pod-os-rejection-status")
+	f.NamespacePodSecurityLevel = admissionapi.LevelBaseline
+	ginkgo.Context("Kubelet", func() {
+		ginkgo.It("should reject pod when the node OS doesn't match pod's OS", func(ctx context.Context) {
+			linuxNode, err := findLinuxNode(ctx, f)
+			framework.ExpectNoError(err)
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "wrong-pod-os",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					OS: &v1.PodOS{
+						Name: "windows", // explicitly set the pod OS to a wrong but valid value
+					},
+					Containers: []v1.Container{
+						{
+							Name:  "wrong-pod-os",
+							Image: imageutils.GetPauseImageName(),
+						},
+					},
+					NodeName: linuxNode.Name, // Set the node to an node which doesn't support
+				},
+			}
+
+			pod = e2epod.NewPodClient(f).Create(ctx, pod)
+			// Check the pod is still not running
+			err = e2epod.WaitForPodFailedReason(ctx, f.ClientSet, pod, "PodOSNotSupported", f.Timeouts.PodStartShort)
+			framework.ExpectNoError(err)
+
+			gotPod, err := f.ClientSet.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			// Check the pod status is set correctly
+			expectedStatus := pod.Status.DeepCopy()
```

Comment:
as commented above, you want to compare status before kubelet updated it and after.

As written, I don't believe this test will even fail with or without the fix you made in kubelet.go

Reply:
@SergeyKanzhelev I test it locally without my fix, the test fails. 

```shell
(base) ➜  kubernetes git:(fix-126662-kubelet) kubectl version
Client Version: v1.31.0
Kustomize Version: v5.4.2
Server Version: v1.31.0

(base) ➜  kubernetes git:(fix-126662-kubelet) export KUBECONFIG=/Users/kiki/.kube/config
(base) ➜  kubernetes git:(fix-126662-kubelet) ginkgo --focus "PodOSRejectionStatus" -v ./test/e2e
[sig-node] PodOSRejectionStatus Kubelet should reject pod when the node OS doesn't match pod's OS [sig-node]
/Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/common/node/pod_admission.go:73
...
• [FAILED] [2.753 seconds]
[sig-node] PodOSRejectionStatus Kubelet [It] should reject pod when the node OS doesn't match pod's OS [sig-node]
/Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/common/node/pod_admission.go:73

  [FAILED] Value for field 'Status' failed to satisfy matcher.
  Expected
      <v1.PodStatus>:
          message: 'Pod was rejected: Failed to admit pod as the OS field doesn''t match node
            OS'
          phase: Failed
          reason: PodOSNotSupported
          startTime: "2024-10-17T07:21:27Z"
  to equal
      <*v1.PodStatus | 0x140027e8000>:
          message: 'Pod was rejected: Failed to admit pod as the OS field doesn''t match node
            OS'
          phase: Failed
          qosClass: BestEffort
          reason: PodOSNotSupported
  In [It] at: /Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/common/node/pod_admission.go:107 @ 10/17/24 15:21:29.793
------------------------------
...
------------------------------
[SynchronizedAfterSuite]
/Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/e2e.go:80
  I1017 15:21:30.351447 6557 suites.go:34] Running AfterSuite actions on node 1
[SynchronizedAfterSuite] PASSED [0.000 seconds]
------------------------------
[ReportAfterSuite] Kubernetes e2e suite report
/Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/e2e_test.go:158
[ReportAfterSuite] PASSED [0.000 seconds]
------------------------------

Summarizing 1 Failure:
  [FAIL] [sig-node] PodOSRejectionStatus Kubelet [It] should reject pod when the node OS doesn't match pod's OS [sig-node]
  /Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/common/node/pod_admission.go:107

Ran 1 of 6603 Specs in 3.064 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 6602 Skipped
```

Reply:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128083/pull-kubernetes-e2e-kind/1846808691345461248

Why it fails in ci? The reject function will set `startTime` field but I forget to mask the `startTime` in the e2e test.

Reply:
Test it locally again without my fix. The startTime filed is the same now.
```
  [FAILED] Value for field 'Status' failed to satisfy matcher.
  Expected
      <v1.PodStatus>:
          message: 'Pod was rejected: Failed to admit pod as the OS field doesn''t match node
            OS'
          phase: Failed
          reason: PodOSNotSupported
          startTime: "2024-10-17T07:53:39Z"
  to equal
      <*v1.PodStatus | 0x14000ac12c0>:
          message: 'Pod was rejected: Failed to admit pod as the OS field doesn''t match node
            OS'
          phase: Failed
          qosClass: BestEffort
          reason: PodOSNotSupported
          startTime: "2024-10-17T07:53:39Z"
  In [It] at: /Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/common/node/pod_admission.go:108
```

Reply:
Test it locally again with my fix.

```
[sig-node] PodOSRejectionStatus Kubelet should reject pod when the node OS doesn't match pod's OS [sig-node]
/Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/common/node/pod_admission.go:73
...
------------------------------
[SynchronizedAfterSuite]
/Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/e2e.go:80
  I1017 16:18:14.236361 26953 suites.go:34] Running AfterSuite actions on node 1
[SynchronizedAfterSuite] PASSED [0.000 seconds]
------------------------------
[ReportAfterSuite] Kubernetes e2e suite report
/Users/kiki/workspace/golang/src/k8s.io/kubernetes/test/e2e/e2e_test.go:158
[ReportAfterSuite] PASSED [0.000 seconds]
------------------------------

Ran 1 of 6603 Specs in 2.332 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 6602 Skipped
PASS
```